compiler/analyzer.ml

```diff
@@ -359,18 +394,65 @@ and check_list env l =
+   * to allow recursion *)
+  let unconst_func_type = Func({
+    param_types = List.map (fun _ -> Unconst) f.params;
+    return_type = Unconst;
+  }) in 
+  let env', name = add_to_scope env id unconst_func_type in
 
   (* Evaluate parameters, body, and return statement in local environment *)
   let func_env, param_ssids = check_fdecl_params env' f.params in
   let func_env, body = check_stmts func_env f.body in
   let func_env, return = check_expr func_env f.return in
-  let Sast.Expr(_, ret_type) = return in
 
+  (* Evaluate parameter and function types *)
+  let rec check_params_type_mismatch env acc func_param_types = function
+    | [] -> env, List.rev acc
+    | ssid :: tl -> 
+        let var = VarMap.find (id_of_ssid ssid) func_env.params and
+        func_param_type = List.hd func_param_types in  
+        (* If both the param var and func_param_type remain unconstrained, 
+         * this parameter can be of any type *)
+        if var.s_type = Unconst && func_param_type = Unconst then 
+          check_params_type_mismatch env (Any :: acc) (List.tl func_param_types) tl
+        (* If the param var and func_param_type are both not Unconst and they 
+         * are the same type, then everything is OK *)
+        else if var.s_type = func_param_type then
+          check_params_type_mismatch env (func_param_type :: acc) (List.tl func_param_types) tl
+        (* If the param var's type is constrained, but the func_param_type
+         * remains unconstrained, constrain the func_param_type *)
+        else if func_param_type = Unconst then
+          check_params_type_mismatch env (var.s_type :: acc) (List.tl func_param_types) tl
+        (* If the func_param_type is constrained, but the func var's type
+         * remains unconstrained, constrain the var's type *)
+        else if var.s_type = Unconst then
+          let env' = constrain_e func_env (Sast.Id(ssid)) func_param_type in 
+          check_params_type_mismatch env' (func_param_type :: acc) (List.tl func_param_types) tl
+        (* If the param var's type does not equal the func_param_type, then 
+         * there is a type mismatch error *)
+        else if var.s_type <> func_param_type then 
+          recursive_type_mismatch_error id var.s_type func_param_type
+        (* TEMPORARY CONDITION FOR TESTING - should never be reached *)
+        else
+          failwith "check_params_type_mismatch in Analyzer is not working correctly" in
+  let param_types = 
+    let typ = (VarMap.find id func_env.scope).s_type in
+    match typ with
+      | Func(func) -> func.param_types
+      | _ -> failwith "check_params_type_mismatch in Analyzer is not working correctly" in
+  let func_env, param_types' = check_params_type_mismatch func_env [] param_types param_ssids in
+  
+  (* Re-evaluate function return type to see if it has been constrained above *)
+  let func_env, return = check_expr func_env f.return in
```

Comment:
right now we're doing it above and again after constraining the params, I think we can wait to evaluate the return type until pretty much the very end (so only call it once down here)